main/src/com/google/refine/model/recon/StandardReconConfig.java

```diff
@@ -460,48 +455,58 @@ public List<Recon> batchRecon(List<ReconJob> jobs, long historyEntryID) {
-                logger.error("Failed to parse string as JSON: " + responseString);
-            } else {
-                for (int i = 0; i < jobs.size(); i++) {
-                    StandardReconJob job = (StandardReconJob) jobs.get(i);
-                    Recon recon = null;
-
-                    String text = job.text;
-                    String key = "q" + i;
-                    if (o.has(key) && o.get(key) instanceof ObjectNode) {
-                        ObjectNode o2 = (ObjectNode) o.get(key);
-                        if (o2.has("result") && o2.get("result") instanceof ArrayNode) {
-                            ArrayNode results = (ArrayNode) o2.get("result");
-
-                            recon = createReconServiceResults(text, results, historyEntryID);
-                        } else {
-                            recon = new Recon(historyEntryID, identifierSpace, schemaSpace);
-                            recon.error = o2.toString();
-                        }
+
+            responseString = postQueries(service, queriesString);
+            o = ParsingUtilities.mapper.readValue(responseString, ObjectNode.class);
+        } catch (IOException e) {
+            Recon recon = new Recon(historyEntryID, identifierSpace, schemaSpace);
+            recon.error = e.getMessage();
+            recons.add(recon);
+        }
+
+        if (o == null) { // utility method returns null instead of throwing
+            Recon recon = new Recon(historyEntryID, identifierSpace, schemaSpace);
+            recon.error = "The reconciliation service returned an invalid response";
+            recons.add(recon);
+
+        } else {
+            for (int i = 0; i < jobs.size(); i++) {
+                StandardReconJob job = (StandardReconJob) jobs.get(i);
+                Recon recon = null;
+
+                String text = job.text;
+                String key = "q" + i;
+                if (o.has(key) && o.get(key) instanceof ObjectNode) {
+                    ObjectNode o2 = (ObjectNode) o.get(key);
+                    if (o2.has("result") && o2.get("result") instanceof ArrayNode) {
+                        ArrayNode results = (ArrayNode) o2.get("result");
+
+                        recon = createReconServiceResults(text, results, historyEntryID);
                     } else {
                         recon = new Recon(historyEntryID, identifierSpace, schemaSpace);
-                        recon.error = o.toString();
-                    }
+                        recon.error = "The service returned a JSON response without \"result\" field for query " + key;
 
-                    if (recon != null) {
-                        recon.service = service;
                     }
-                    recons.add(recon);
+                } else {
+                    recon = new Recon(historyEntryID, identifierSpace, schemaSpace);
+                    recon.error = "The service returned a JSON response without \"key\" field ";
```

Comment:
This error message is a bit misleading: we are not expecting a `"key"` field, but rather a field whose name is equal to the contents of the `key` Java variable. So it should rather be something like:
```suggestion
                    recon.error = "The service returned a JSON response without \"" + key + "\" field ";
```